fix(types): enable TypeScript libCheck & resolve type conflicts

davidgovea opened the pull request:
### Related issues
Fixes #2614 
Also see #2732 , #3728

### Description

I :heart: `skipLibCheck: false`! The issues it catches are often non-trivial.

This PR changes the following:
1. Updates the TypeScript definition structure to allow `skipLibCheck: false` compatibility
   * Avoids errors like:
     * `Import declaration conflicts with local declaration of 'ReactNativeFirebase'`
     * `Exports and export assignments are not permitted in module augmentations.`
     * `Top-level declarations in .d.ts files must start with either a 'declare' or 'export' modifier.`
1. Enabling the "libCheck" also uncovered some issues with the existing types, so these issues have been fixed with a **minimal-change** approach.
   * For example, the `SetCheckoutOptionEventParameters` interface was missing, so it acted as `any`.
     * This PR changes the type to `any`, rather than adding the missing interface.
     * I will submit a separate PR with type improvements. Minimal-change approach here.

---

### Approach:

Previously, module `.d.ts` files had the following:
1. `declare module '@react-native-firebase/MODULE'` with the module's exports
2. `declare module '@react-native-firebase/app'`, augmenting the root `firebase` object
3. (Optional) a `namespace ReactNativeFirebase` to augment the `FirebaseJsonConfig` interface.

This PR switches (1) to root-level exports, leaves (2) as-is, and merges (3) into (2) to properly merge the namespace, rather than override it.

For the type fixes, each was placed into a separate commit for easier review. Can definitely squash if desired.. there are a handful!

### Checklist

- I read the [Contributor Guide](../CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
  - [x] Yes
- My change supports the following platforms;
  - [x] `Android`
  - [x] `iOS`
- My change includes tests;
  - [ ] `e2e` tests added or updated in `packages/\*\*/e2e` [**No code changes**]
  - [x] `jest` tests added or updated in `packages/\*\*/__tests__`
- [x] I have updated TypeScript types that are affected by my change.
- This is a breaking change;
  - [ ] Yes
  - [x] No



### Test Plan

Type tests are passing, and one intentionally-incorrect test was updated with a type override, since TS now catches the error.

Test patches to be used with `patch-package` https://gist.github.com/davidgovea/2feea6af41b41f76fae1a9c14465cb69
---

Thanks for reading! I know that @mikehardy has probably already read this before I submitted it.. somehow.. :mage:

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/invertase/react-native-firebase?pullRequest=4306) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/invertase/react-native-firebase?pullRequest=4306) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/invertase/react-native-firebase?pullRequest=4306) it.</sub>

davidgovea commented:
Here is a gist of each package's changes in patch files for use with `patch-package`

https://gist.github.com/davidgovea/2feea6af41b41f76fae1a9c14465cb69

davidgovea commented:
@mikehardy It looks like the automatic patch creation is having issues: patches were only generated for `app`, `analytics`, and `admob` modules.

The others have errors like: `Error: Can't find lockfile entry for @react-native-firebase/ml-vision`

mikehardy commented:
:facepalm: thanks for pointing that out - it's relatively new but I swear it was working when I tested it. Back to the workbench!

davidgovea commented:
Made an improvement to the `ml-vision` types -- I had previously dropped down to `any` on some of the MLKit types that were not found, but I noticed they existed in the `BarcodeDetectorTypes.d.ts` file, so I imported them. It also uncovered a typo in the type-test for ml-vision.

Updated the patches in my gist as well.

Salakar commented:
Looks like theres an issue with typedocs command failing with these changes which is causing the vercel deploy preview failure:

![image](https://user-images.githubusercontent.com/5347038/94398874-4c161000-015e-11eb-928e-ba3b4e650cf1.png)

Salakar commented:
Side note we should probably run this script https://github.com/invertase/react-native-firebase/blob/master/website/scripts/generate-typedoc.js as part of the `
Code Quality Checks / TypeScript Build Validation (pull_request)` CI stage

davidgovea commented:
Thanks for reviewing @Salakar!

This is popping up because of the older version of TypeScript being used in this portion of the project (3.5.3) vs the root project's `^3.8.3`.

TS improved recursive types a lot in 3.7 (see https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#more-recursive-type-aliases and https://github.com/microsoft/TypeScript/pull/33050), and they make a lot of sense for the Firestore definitions, so it seems to me that upgrading typedoc is appropriate. I'll make an attempt.

Another TS feature that I wished was available for the `type-test`s is the `// @ts-expect-error` comment from 3.9. Would allow great improvements to the type testing in an easy way.

davidgovea commented:
Well, that didn't go as planned.

Typedoc is currently locked to `0.15.0`, 0.16 updates to TS >3.7, **but** it also changes some internal behavior. The [`inputFiles`](https://github.com/invertase/react-native-firebase/blob/master/website/scripts/generate-typedoc.js#L28) are empty when using a more recent version of typedoc (tried latest 0.19.2 and 0.16.0).

I'll be looking into it more when I have some more time to circle back.

mikehardy commented:
Honest question: why not bump to typescript 4.x? I'm not aware of compatibility issues (I just bump typescript whenever it comes out since it has never caused me problems, and RNFB is fully working for me with typescript "current stable" (4.x whatever it is right now)

As a developer dependency that is just an optional support for the package I'd be +1 on bumping typescript all over. Historically I believe we have treated type breaks as a minor version ? if it is clearly documented we bumped minimum typescript version that would be okay with me but I'm also in the "version numbers are free" school of thought so could do semver major if needed (with same need for a clear breaking change / exact step to migrate snip in the changelog - easy to do)

davidgovea commented:
Alrighty -- the website is building properly.

I don't have access to the vercel action info, but I have tested locally and things are looking good.

Updated to the latest version of Typedoc, and added the one required change: a new `TSConfigReader` reader plugin call.

---

@mikehardy -- I agree. I actually did bump TS to v4 in another side-branch, and things seemed fine. Just didn't want to overload this PR with extras.
Also, it looks like typedoc has not moved to v4 yet.

mikehardy commented on packages/messaging/lib/index.d.ts:
```diff
@@ -608,7 +608,7 @@ export namespace FirebaseMessagingTypes {
      * @param scope The scope to assign a token, which the sever can use to target messages at.
      */
-    getToken(authorizedEntity?: string, scope?: string = 'FCM'): Promise<string>;
+    getToken(authorizedEntity?: string, scope?: string): Promise<string>;
```
Interesting - this default string value you are removing has some relation to this issue: #3714

davidgovea replied:
Hey @mikehardy -- the place that the default `scope = 'FCM'` is set is in the JS code here: https://github.com/invertase/react-native-firebase/blob/master/packages/messaging/lib/index.js#L142-L159.

`.d.ts` definitions have no runtime responsibilities: parameter defaults are not allowed.

If changing the default scope is a change needed in #3714, it needs to happen in the source file.

mikehardy replied:
Yes definitely - I just thought it was interesting the 'FCM' scoped was baked in here as well, and since it's an active discussion item I was cross-linking for cross-pollination. I commented over there that I'd love a PR to do the actual change

mikehardy reviewed (commented):
Just a note I scanned through these again they all seem reasonable. Going to test the patches now in my work project since otherwise the type checks here aren't very robust I don't think. I'm mostly interested in whether these cause any breaking changes or not, for correct callouts in release notes

mikehardy reviewed (approved):
I integrated all the patches in my work project and everything seems fine, no type breaks there. I'm not able to set skipLibCheck to false in my project due to other problems, but none of them were in react-native-firebase after this change
